libraries/bot-ai-qna/src/main/java/com/microsoft/bot/ai/qna/QnAMaker.java

```diff
@@ -0,0 +1,338 @@
+     *                            the QnaMessage event.
+     * @return A list of answers for the user query, sorted in decreasing order of ranking score.
+     */
+    public CompletableFuture<QueryResult[]> getAnswers(TurnContext turnContext, QnAMakerOptions options,
+                                                       Map<String, String> telemetryProperties,
+                                                       @Nullable Map<String, Double> telemetryMetrics) {
+        return this.getAnswersRaw(turnContext, options, telemetryProperties, telemetryMetrics)
+            .thenApply(result -> result.getAnswers());
+    }
+
+    /**
+     * Generates an answer from the knowledge base.
+     * @param turnContext The Turn Context that contains the user question to be queried against your knowledge base.
+     * @param options The options for the QnA Maker knowledge base.
+     *                If null, constructor option is used for this instance.
+     * @param telemetryProperties Additional properties to be logged to telemetry with the QnaMessage event.
+     * @param telemetryMetrics Additional metrics to be logged to telemetry with the QnaMessage event.
+     * @return A list of answers for the user query, sorted in decreasing order of ranking score.
+     */
+    public CompletableFuture<QueryResults> getAnswersRaw(TurnContext turnContext, QnAMakerOptions options,
+                                                              Map<String, String> telemetryProperties,
+                                                              @Nullable Map<String, Double> telemetryMetrics) {
+        if (turnContext == null) {
+            throw new IllegalArgumentException("turnContext");
+        }
+        if (turnContext.getActivity() == null) {
+            throw new IllegalArgumentException(String.format("The %1$s property for %2$s can't be null.",
+                "Activity", "turnContext"));
+        }
+        Activity messageActivity = turnContext.getActivity();
+        if (messageActivity == null || messageActivity.getType() != ActivityTypes.MESSAGE) {
+            throw new IllegalArgumentException("Activity type is not a message");
+        }
+        if (Strings.isNullOrEmpty(turnContext.getActivity().getText())) {
+            throw new IllegalArgumentException("Null or empty text");
+        }
+
+        return this.generateAnswerHelper.getAnswersRaw(turnContext, messageActivity, options).thenApply(result -> {
+            onQnaResult(result.getAnswers(), turnContext, telemetryProperties, telemetryMetrics);
+            return result;
+        });
+    }
+
+    /**
+     * Filters the ambiguous question for active learning.
+     * @param queryResult User query output.
+     * @return Filtered array of ambiguous question.
+     */
+    public QueryResult[] getLowScoreVariation(QueryResult[] queryResult) {
+        return ActiveLearningUtils.getLowScoreVariation(Arrays.asList(queryResult)).toArray(queryResult);
```

Comment:
Why `queryResult` inside the `toArray`?

Reply:
If it's because you need to parse it I think that it's better the following way:
`return (QueryResult[]) ActiveLearningUtils.getLowScoreVariation(Arrays.asList(queryResult)).toArray();`